File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.js
const input = require("fs")
  .readFileSync(0)
  .toString()
  .trim()
  .split("\n")
  .map((v) => v.trim());

const N = Number(input[0]);
const NUMS = input[1].split(" ").map(Number);
const calcs = input[2].split(" ").map(Number);
const CALCS = [];

for (let i = 0; i < 4; i++) {
  if (i === 0) CALCS.push(..."+".repeat(calcs[i]).split(""));
  if (i === 1) CALCS.push(..."-".repeat(calcs[i]).split(""));
  if (i === 2) CALCS.push(..."*".repeat(calcs[i]).split(""));
  if (i === 3) CALCS.push(..."/".repeat(calcs[i]).split(""));
}

const permutations = (arr, n) => {
  if (n === 1) return [[]];

  const result = [];
  arr.forEach((fixed, idx) => {
    const rest = [...arr.slice(0, idx), ...arr.slice(idx + 1)];
    const perms = permutations(rest, n - 1);
    const combine = perms.map((p) => [fixed, ...p]);
    result.push(...combine);
  });

  return result;
};

const calculate = (a, operator, b) => {
  if (operator === "+") return a + b;
  if (operator === "-") return a - b;
  if (operator === "*") return a * b;
  if (operator === "/") return a < 0 ? -Math.floor(-a / b) : Math.floor(a / b);
};

let max = -Infinity;
let min = Infinity;

const backtrack = (depth, value, usedOperators) => {
  if (depth === N - 1) {
    max = Math.max(max, value);
    min = Math.min(min, value);
    return;
  }

  for (let i = 0; i < CALCS.length; i++) {
    if (!usedOperators[i]) {
      usedOperators[i] = true;
      const nextValue = calculate(value, CALCS[i], NUMS[depth + 1]);
      backtrack(depth + 1, nextValue, usedOperators);
      usedOperators[i] = false;
    }
  }
};

backtrack(0, NUMS[0], Array(CALCS.length).fill(false));

console.log([max, min].join("\n"));

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
import sys
input = sys.stdin.readline

from itertools import permutations

N = int(input())
NUMS = list(map(int, input().split()))
calcs = list(map(int, input().split()))
CALCS = []

for i in range(4):
   if i == 0: 
      CALCS.extend(['+'] * calcs[i])
   if i == 1: 
      CALCS.extend(['-'] * calcs[i])
   if i == 2: 
      CALCS.extend(['*'] * calcs[i])
   if i == 3: 
      CALCS.extend(['/'] * calcs[i])

def calculate(a, operator, b):
   if(operator == '+'):
      return a + b
   if(operator == '-'):
      return a - b
   if(operator == '*'):
      return a * b
   if(operator == '/'):
      if a < 0 :
         return -(-a // b)
      else:
         return a // b
      
max_value = float('-inf')
min_value = float('inf')

used_operators = [False] * (N - 1)

def backtrack(depth, value):
   global max_value, min_value
   
   if depth == N - 1:
      max_value = max(max_value, value)
      min_value = min(min_value, value)
      return
   
   for i in range(len(CALCS)):
      if not used_operators[i]:
         used_operators[i] = True
         next_value = calculate(value, CALCS[i], NUMS[depth + 1])
         backtrack(depth + 1, next_value)
         used_operators[i] = False
         
backtrack(0, NUMS[0])

print("\n".join(map(str, [max_value, min_value])))